### PR TITLE
[WIP]: Fulltext search (#8381)

### PR DESCRIPTION
**Pull request by mrsimpson**
@geekgonecrazy this PR is only used to discuss WIP of search interface and sample implementation
Addresses #8381

### CODE REVIEW

**Comment by mrsimpson** [comment]
@tkurz: Please check interface definition. Can you start from here?

**Comment by geekgonecrazy** [comment]
@mrsimpson can you base these changes off of develop?  The change list here is massive.  Hard to collaborate, get feedback wading through all of those files.  :grin:  The only files that should matter are the ones to the search package you just added here.

**Comment by graywolf336** [comment]
I agree with @geekgonecrazy, the change list is hard to follow...

File: packages/rocketchat-search-lib/lib/SearchProviderRegistry.js [hunk 1]
```diff
@@ -1,13 +1,33 @@
...
+	add(searchProvider) {
+		if (!(searchProvider instanceof SearchProvider)) {
+			throw new Error('Invalid provider object, it must extend "SearchProvider"');
+		}
+
+		if (this.providers[searchProvider.identifier]) {
+			return false;
+		}
+
+		this.providers[searchProvider.identifier] = searchProvider;
```
Comment (graywolf336) [inline packages/rocketchat-search-lib/lib/SearchProviderRegistry.js]: Since the documentation on this method says it returns a boolean, make sure you return one at the end of it. :)

**Comment by mrsimpson** [comment]
I’ll do a rebase soon, that will make it easier. Was in a hurry yday. It’s actually currently only two packages.

File: packages/rocketchat-search-lib/lib/SearchProviderRegistry.js (hunk 1 above)
Comment (mrsimpson) [reply packages/rocketchat-search-lib/lib/SearchProviderRegistry.js]: Of course. Thx. There's basically no implementation yet. Intention of sharing it was to be able to discuss about the interfaces. I might add a bit more during the next days.

**Comment by mrsimpson** [comment]
@geekgonecrazy @graywolf336 merged `develop`. Didn't notice the branch-off-point before I pushed it. Sorry for that!
I would really try to stick to those two packages. I plan to integrate via webhooks which then in their implementation trigger the active search provider.
Thus, I'd only add hooks if needed.
With respect to the UI I am uncertain. I don't like the integration in the tabbar, as this one currently serves only purposes in the context of the room - and search is cross-room.
Therefore, I'd go and integrate into spotlight. We'll check which UI suits us. Ok?

**Comment by geekgonecrazy** [comment]
> With respect to the UI I am uncertain. I don't like the integration in the tabbar, as this one currently serves only purposes in the context of the room - and search is cross-room.
Therefore, I'd go and integrate into spotlight. We'll check which UI suits us. Ok?

Regarding UI I know the sidebar is changing.  I think search will move to the top and it may become a bit more fitting for global.

But at any rate I think for the scope of implementing search providers and search provider api's we should provide methods to be called to perform the search, but we should not go so far as to hook into the UI. 

This way the frontend guys can work with designers to get that all figured out.

File: packages/redlink-search/lib/SearchProvider.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+		}
+	}
+}
+
+/*
+Don't know whether an own UI is really necessary
+ */
+class RedlinkSearchProviderUi extends SearchProviderUi {
+
+}
```
Comment (geekgonecrazy) [inline packages/redlink-search/lib/SearchProvider.js]: If implemented what would this do beyond the settings exposed?

Comment (mrsimpson) [reply packages/redlink-search/lib/SearchProvider.js]: @geekgonecrazy there are usecases where a search provider has own capabilities, such as filtering, facetting, more-like-this, ... 
A generic UI would have to be highly configurable in order to deal with all the different options - and it would be tricky to get all providers aligned.

Comment (geekgonecrazy) [reply packages/redlink-search/lib/SearchProvider.js]: @graywolf336 thoughts? :top: this is going to be important across various parts of rocketchat apps.

Also we don't want to count on blaze.... as we might be switching frameworks.  So something agnostic would be best if this is needed.

Comment (mrsimpson) [reply packages/redlink-search/lib/SearchProvider.js]: @geekgonecrazy one more thing about that: many external systems (which are to be integrated) already bring some sort of UI. For search, there are plenty of libs. None of them is written in blaze. This, a template which loads the other technology‘s UI onRendered is at least something to consider. As you said: also for the RC apps

**Comment by geekgonecrazy** [comment]
@mrsimpson do we want to close this in favor of #10110 ?

**Comment by mrsimpson** [comment]
Of course! This one is kind-of the forerunner ;) Looking so much forward to seeing #10110 merged!
